core/templates/pages/exploration-player-page/services/stats-reporting.service.ts

```diff
@@ -16,344 +16,416 @@
  * @fileoverview Services for stats reporting.
  */
 
-require('domain/utilities/StopwatchObjectFactory.ts');
-require('domain/utilities/url-interpolation.service.ts');
-require(
-  'pages/exploration-player-page/services/answer-classification.service.ts');
-require('services/context.service.ts');
-require('services/messenger.service.ts');
-require('services/playthrough.service.ts');
-require('services/site-analytics.service.ts');
-
-require(
-  'pages/exploration-player-page/exploration-player-page.constants.ajs.ts');
-
-angular.module('oppia').factory('StatsReportingService', [
-  '$http', '$interval', 'ContextService', 'MessengerService',
-  'PlaythroughService', 'SiteAnalyticsService', 'StopwatchObjectFactory',
-  'UrlInterpolationService', 'STATS_REPORTING_URLS',
-  function(
-      $http, $interval, ContextService, MessengerService,
-      PlaythroughService, SiteAnalyticsService, StopwatchObjectFactory,
-      UrlInterpolationService, STATS_REPORTING_URLS) {
-    var explorationId = null;
-    var explorationTitle = null;
-    var explorationVersion = null;
-    var sessionId = null;
-    var stateStopwatch = null;
-    var optionalCollectionId = undefined;
-    var statesVisited = {};
-    var numStatesVisited = 0;
-    var explorationStarted = false;
-    var explorationActuallyStarted = false;
-    var explorationIsComplete = false;
-    var currentStateName = null;
-    var nextExpId = null;
-    var previousStateName = null;
-    var nextStateName = null;
-
-    var _editorPreviewMode = ContextService.isInExplorationEditorPage();
-    var _questionPlayerMode = ContextService.isInQuestionPlayerMode();
+import { downgradeInjectable } from '@angular/upgrade/static';
+import { HttpClient } from '@angular/common/http';
+import { Injectable } from '@angular/core';
+
+import { ContextService } from 'services/context.service';
+import { ExplorationPlayerConstants } from
+  'pages/exploration-player-page/exploration-player-page.constants';
+import { MessengerService } from 'services/messenger.service';
+import { PlaythroughService } from 'services/playthrough.service';
+import { SiteAnalyticsService } from 'services/site-analytics.service';
+import { Stopwatch, StopwatchObjectFactory } from
+  'domain/utilities/StopwatchObjectFactory';
+import { UrlInterpolationService } from
+  'domain/utilities/url-interpolation.service';
+
+interface IStateStats {
```

Comment:
Hmm, let's actually keep `stateName` out of the interface:

```typescript
interface IStateStats {
  /* eslint-disable camelcase */
  total_answers_count: number;
  useful_feedback_count: number;
  total_hit_count: number;
  first_hit_count: number;
  num_times_solution_viewed: number;
  num_completions: number;
  /* eslint-enable camelcase */
}

interface IAggregatedStats {
  /* eslint-disable camelcase */
  num_starts: number;
  num_completions: number;
  num_actual_starts: number;
  state_stats_mapping: {[stateName: string]: IStateStats};
  /* eslint-enable camelcase */
}

Reply:
Done.